File: stepik-union/src/main/java/org/stepik/plugin/actions/navigation/StepikStepNavigationAction.java
package org.stepik.plugin.actions.navigation;

import com.intellij.openapi.project.Project;
import com.jetbrains.tmp.learning.StepikProjectManager;
import com.jetbrains.tmp.learning.courseFormat.StudyNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.plugin.utils.NavigationUtils;

import javax.swing.*;

abstract class StepikStepNavigationAction extends StudyStepNavigationAction {
    StepikStepNavigationAction(@Nullable String text, @Nullable String description, @Nullable Icon icon) {
        super(text, description, icon);
    }

    @Override
    public void navigateStep(@NotNull final Project project) {
        StudyNode currentNode = StepikProjectManager.getSelected(project);
        if (currentNode == null) {
            currentNode = StepikProjectManager.getProjectRoot(project);
        }

        StudyNode targetNode;

        targetNode = getTargetStep(currentNode);

        if (targetNode == null) {
            return;
        }

        NavigationUtils.navigate(project, targetNode);
    }
}

File: stepik-union/src/main/java/org/stepik/core/actions/navigation/StepikStepNavigationAction.kt
package org.stepik.core.actions.navigation

import com.intellij.openapi.project.Project
import org.stepik.core.StepikProjectManager
import org.stepik.core.courseFormat.StudyNode
import org.stepik.core.utils.NavigationUtils
import javax.swing.Icon

abstract class StepikStepNavigationAction protected constructor(text: String?,
                                                                description: String?,
                                                                icon: Icon?) :
        StudyStepNavigationAction(text, description, icon) {

    public override fun navigateStep(project: Project) {
        var currentNode = StepikProjectManager.getSelected(project)
        if (currentNode == null) {
            currentNode = StepikProjectManager.getProjectRoot(project)
        }

        val targetNode: StudyNode<*, *>?

        targetNode = getTargetStep(currentNode)

        if (targetNode == null) {
            return
        }

        NavigationUtils.navigate(project, targetNode)
    }
}
